File: ganttproject/src/main/java/net/sourceforge/ganttproject/gui/FileChooserPageBase.java
/*
GanttProject is an opensource project management tool.
Copyright (C) 2011 GanttProject team

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 3
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
package net.sourceforge.ganttproject.gui;

import biz.ganttproject.app.InternationalizationKt;
import biz.ganttproject.core.option.BooleanOption;
import biz.ganttproject.core.option.DefaultBooleanOption;
import biz.ganttproject.core.option.GPOptionGroup;
import javafx.beans.property.SimpleObjectProperty;
import net.sourceforge.ganttproject.document.Document;
import net.sourceforge.ganttproject.gui.options.OptionsPageBuilder;
import net.sourceforge.ganttproject.gui.projectwizard.WizardPage;
import org.eclipse.core.runtime.IStatus;
import org.eclipse.core.runtime.Status;
import org.osgi.service.prefs.Preferences;

import javax.swing.*;
import javax.swing.filechooser.FileFilter;
import java.awt.*;
import java.io.File;
import java.net.MalformedURLException;
import java.net.URL;

public abstract class FileChooserPageBase implements WizardPage {
  protected static final String PREF_SELECTED_FILE = "selected_file";
  private final Document myDocument;

  private final TextFieldAndFileChooserComponent myChooser;
  private final OptionsPageBuilder myOptionsBuilder;
  private final JPanel mySecondaryOptionsComponent;
  private final JLabel myFileLabel = new JLabel("");
  private final BooleanOption myOverwriteOption = new DefaultBooleanOption("overwrite");

  private final Preferences myPreferences;

  public SimpleObjectProperty selectedUrlProperty = new SimpleObjectProperty<URL>();
  protected FileChooserPageBase(Preferences prefs, Document document, UIFacade uiFacade) {
    myPreferences = prefs;
    myDocument = document;
    myOptionsBuilder = new OptionsPageBuilder();
    mySecondaryOptionsComponent = new JPanel(new BorderLayout());
    mySecondaryOptionsComponent.setBorder(BorderFactory.createEmptyBorder(10, 0, 0, 0));
    myOptionsBuilder.setI18N(new OptionsPageBuilder.I18N() {
      @Override
      protected boolean hasValue(String key) {
        return (key.equals(getCanonicalOptionLabelKey(myOverwriteOption) + ".trailing")) ? true : super.hasValue(key);
      }

      @Override
      protected String getValue(String key) {
        if (key.equals(getCanonicalOptionLabelKey(myOverwriteOption) + ".trailing")) {
          return InternationalizationKt.getRootLocalizer().formatText("document.overwrite");
        }
        return super.getValue(key);
      }
    });
    myChooser = new TextFieldAndFileChooserComponent(uiFacade, getFileChooserTitle()) {
      @Override
      protected void onFileChosen(File file) {
        tryChosenFile(file);
      }
    };
  }

  protected abstract String getFileChooserTitle();

  protected BooleanOption getOverwriteOption() {
    return myOverwriteOption;
  }
  /** @return a default export filename */
  protected String getDefaultFileName() {
    if (myDocument == null) {
      return "document.gan";
    }
    return myDocument.getFileName();
  }

  protected int getFileChooserSelectionMode() {
    return JFileChooser.FILES_AND_DIRECTORIES;
  }

  protected void tryChosenFile(File file) {
    IStatus status = FileChooserPageBase.this.onSelectedFileChange(file);
    if (!status.isOK()) {
      onSelectedUrlChange(null);
    }
    FileChooserPageBase.setStatus(myFileLabel, status);
  }

  @Override
  public Component getComponent() {
      JPanel myComponent = new JPanel(new BorderLayout());
    myChooser.setFileSelectionMode(getFileChooserSelectionMode());
    JComponent contentPanel = new JPanel(new BorderLayout());
    Box fileBox = Box.createVerticalBox();
    myChooser.setAlignmentX(Component.LEFT_ALIGNMENT);
    fileBox.add(myChooser);
    myFileLabel.setAlignmentX(Component.LEFT_ALIGNMENT);
    fileBox.add(myFileLabel);
    fileBox.add(myOptionsBuilder.createOptionComponent(new GPOptionGroup("exporter", myOverwriteOption), myOverwriteOption));
    contentPanel.add(fileBox, BorderLayout.NORTH);
    contentPanel.add(mySecondaryOptionsComponent, BorderLayout.CENTER);
    myComponent.add(contentPanel, BorderLayout.NORTH);
    return myComponent;
  }

  protected void loadPreferences() {
    String oldFile = myPreferences.get(FileChooserPageBase.PREF_SELECTED_FILE, null);
    if (oldFile != null) {
      // Use the previously used path with the current filename for the default
      // name
      // The implementing classes can modify the file extension when desired
      String oldPath = new File(oldFile).getParent();
      File f = new File(oldPath, getDefaultFileName());
      myChooser.setFile(f);
    }
  }

  @Override
  public void setActive(boolean b) {
    GPOptionGroup[] optionGroups = getOptionGroups();
    if (b == false) {
        for (GPOptionGroup optionGroup : optionGroups) {
            optionGroup.commit();
        }
      if (myChooser.getFile() != null) {
        myPreferences.put(FileChooserPageBase.PREF_SELECTED_FILE, myChooser.getFile().getAbsolutePath());
      }
    } else {
        for (GPOptionGroup optionGroup : optionGroups) {
            optionGroup.lock();
        }
      if (mySecondaryOptionsComponent != null) {
        mySecondaryOptionsComponent.removeAll();
      }
      mySecondaryOptionsComponent.add(createSecondaryOptionsPanel(), BorderLayout.NORTH);
      myChooser.setFileFilter(createFileFilter());
      loadPreferences();
      onSelectedUrlChange(getSelectedUrl());
      //myWizard.getDialog().layout();
    }
  }

  protected Component createSecondaryOptionsPanel() {
    return myOptionsBuilder.buildPlanePage(getOptionGroups());
  }

  protected final Preferences getPreferences() {
    return myPreferences;
  }

  protected final TextFieldAndFileChooserComponent getChooser() {
    return myChooser;
  }

  private URL getSelectedUrl() {
     return myChooser.getSelectedURL();
  }

  protected abstract FileFilter createFileFilter();

  protected abstract GPOptionGroup[] getOptionGroups();

  protected void onSelectedUrlChange(URL selectedUrl) {
    selectedUrlProperty.set(selectedUrl);
  }

  protected IStatus setSelectedFile(File file) {
    try {
      onSelectedUrlChange(file.toURI().toURL());
      return new Status(IStatus.OK, "foo", IStatus.OK, "  ", null);
    } catch (MalformedURLException e) {
      e.printStackTrace();
      return new Status(IStatus.ERROR, "foo", IStatus.ERROR, e.getMessage(), null);
    }
  }

  protected IStatus onSelectedFileChange(File file) {
    if (file == null) {
      return new Status(IStatus.ERROR, "foo", IStatus.ERROR, "File does not exist", null);
    }
    if (!file.exists()) {
      return new Status(IStatus.ERROR, "foo", IStatus.ERROR, "File does not exist", null);
    }
    if (!file.canRead()) {
      return new Status(IStatus.ERROR, "foo", IStatus.ERROR, "File read error", null);
    }
    return setSelectedFile(file);
  }

  private static void setStatus(JLabel label, IStatus status) {
    label.setOpaque(true);
    if (status.isOK()) {
      UIUtil.clearErrorLabel(label);
      label.setText(status.getMessage());
    } else {
      UIUtil.setupErrorLabel(label, status.getMessage());
    }
  }
}

File: ganttproject/src/main/java/net/sourceforge/ganttproject/gui/FileChooserPageBase.kt
/*
 * Copyright (c) 2011-2026 Dmitry Barashev, BarD Software s.r.o.
 *
 * This file is part of GanttProject, an open-source project management tool.
 *
 * GanttProject is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 * GanttProject is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with GanttProject.  If not, see <http://www.gnu.org/licenses/>.
 */
package net.sourceforge.ganttproject.gui

import biz.ganttproject.app.*
import biz.ganttproject.core.option.*
import biz.ganttproject.lib.fx.GPListCell
import biz.ganttproject.lib.fx.hbox
import biz.ganttproject.lib.fx.vbox
import com.github.michaelbull.result.Result
import com.github.michaelbull.result.onFailure
import com.github.michaelbull.result.onSuccess
import javafx.collections.FXCollections
import javafx.collections.ObservableList
import javafx.embed.swing.SwingNode
import javafx.scene.Node
import javafx.scene.control.ContentDisplay
import javafx.scene.control.ListView
import javafx.scene.layout.BorderPane
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.javafx.JavaFx
import kotlinx.coroutines.launch
import kotlinx.coroutines.swing.Swing
import kotlinx.coroutines.withContext
import net.sourceforge.ganttproject.action.GPAction
import net.sourceforge.ganttproject.document.Document
import net.sourceforge.ganttproject.gui.options.OptionsPageBuilder
import org.osgi.service.prefs.Preferences
import java.awt.BorderLayout
import java.awt.Component
import java.io.File
import javax.swing.BorderFactory
import javax.swing.JFileChooser
import javax.swing.JPanel
import javax.swing.SwingUtilities

/**
 * Base class for the file chooser pages in the Import and Export wizards.
 */
abstract class FileChooserPageBase protected constructor(
  private val myDocument: Document?,
  val fileChooserTitle: String?,
  val pageTitle: String?,
  val fileChooserSelectionMode: Int = JFileChooser.FILES_ONLY,
  val errorMessage: ObservableString,
  val coroutineScope: CoroutineScope
) : WizardPage {
  val fxFile = ObservableFile("file", null)
  val fxFiles = ObservableFiles("files", emptyList())
  val chosenFiles = FXCollections.observableArrayList<ChosenFile>()
  var allowMultipleChoice: Boolean = false

  private var fileFilter: FileExtensionFilter? = null
    set(value) {
      field = value
      if (value != null) {
        extensionFilters.clear()
        extensionFilters.add(value)
      }
    }
  private var extensionFilters: MutableList<FileExtensionFilter> = mutableListOf()
    set(value) {
      if (field.isNotEmpty()) {
        value.addAll(field)
      }
      field = value
    }
  protected val fxOverwrite = ObservableBoolean("overwrite", false)
  abstract val preferences: Preferences
  private val secondaryOptionsSwingNode = SwingNode()

  init {
    fxOverwrite.addWatcher { tryChosenFile(fxFile.value) }
  }

  override val component: Component? = null
  private var resetCenterPane: ()->Unit = {}

  override val fxComponent: Node by lazy {
    val root = BorderPane()
    root.styleClass.add("file-chooser-page")
    val i18n = i18n {
      default()
      map(mapOf(
        "file.label" to "file",
        "files.label" to "file",
        "overwrite.label" to "option.exporter.overwrite.label.trailing"
      ))
    }
    if (allowMultipleChoice) {
      val updateFileList = { files: List<File> ->
        chosenFiles.clear()
        files.forEach { file ->
          val newChosenFile = ChosenFile(file, fxFiles,chosenFiles, this::validateFile)
          if (!chosenFiles.contains(newChosenFile)) {
            chosenFiles.add(newChosenFile)
          }
        }
      }
      fxFiles.addWatcher { updateFileList(it.newValue) }
      updateFileList(fxFiles.value)

    } else {
      fxFile.addWatcher { event ->
        tryChosenFile(event.newValue)
//        if (allowMultipleChoice) {
//          event.newValue?.let {
//            val newChosenFile = ChosenFile(event.newValue!!, fxFchosenFiles, this::validateFile)
//            if (!chosenFiles.contains(newChosenFile)) {
//              chosenFiles.add(newChosenFile)
//            }
//          }
//        }
      }
    }

    val sheet = PropertySheetBuilder(i18n).pane {
      if (allowMultipleChoice) {
        files(fxFiles) {
          chooserTitle = fileChooserTitle ?: ""
          isSaveNotOpen = fileChooserSelectionMode != JFileChooser.FILES_ONLY
          browseButtonText = RootLocalizer.formatText("fileChooser.browse")
          allowMultipleSelection = true
          editorStyles.add("file-chooser")
          this@FileChooserPageBase.extensionFilters = this.extensionFilters
        }
      } else {
        file(fxFile) {
          chooserTitle = fileChooserTitle ?: ""
          isSaveNotOpen = fileChooserSelectionMode != JFileChooser.FILES_ONLY
          browseButtonText = RootLocalizer.formatText("fileChooser.browse")
          editorStyles.add("file-chooser")
          this@FileChooserPageBase.extensionFilters = this.extensionFilters
        }
      }
      if (hasOverwriteOption) {
        checkbox(fxOverwrite)
      }
    }
    root.top = sheet.node

    resetCenterPane = {
      coroutineScope.launch {
        val addOptionsComponentSwing: Boolean = withContext(Dispatchers.JavaFx) {
          val optionsComponentFx: Node? = createSecondaryOptionsPanelFx()
          root.center = vbox {
            if (allowMultipleChoice) {
              val listView = ListView(chosenFiles).also {
                it.styleClass.addAll("chosen-files-list", "swing-background")
                it.setCellFactory { _ -> ListCellImpl() }
                it.prefHeight = 200.0
              }
              add(listView)
            }
            if (optionsComponentFx == null) {
              add(secondaryOptionsSwingNode)
            } else {
              add(optionsComponentFx)
            }
          }
          return@withContext optionsComponentFx == null
        }
        if (addOptionsComponentSwing) {
          withContext(Dispatchers.Swing) {
            mySecondaryOptionsComponent.removeAll()
            mySecondaryOptionsComponent.add(createSecondaryOptionsPanel(), BorderLayout.NORTH)
            secondaryOptionsSwingNode.content = mySecondaryOptionsComponent
          }
        }
      }
    }
    resetCenterPane()
    val errorPane = ErrorPane()
    root.bottom = errorPane.fxNode

    errorMessage.addWatcher { evt ->
      FXThread.runLater {
        errorPane.onError(evt.newValue?.let(::html2md))
      }
    }
    errorPane.onError(errorMessage.value?.let(::html2md))
    root
  }

  private fun showError(msg: String?) {
    errorMessage.value = msg
  }

  private val myOptionsBuilder: OptionsPageBuilder = OptionsPageBuilder().also {
//    it.i18N = object : OptionsPageBuilder.I18N() {
//      protected override fun hasValue(key: String): Boolean {
//        return if (key == getCanonicalOptionLabelKey(overwriteOption) + ".trailing") true else super.hasValue(
//          key
//        )
//      }
//
//      protected override fun getValue(key: String): String? {
//        if (key == getCanonicalOptionLabelKey(overwriteOption) + ".trailing") {
//          return RootLocalizer.formatText("document.overwrite")
//        }
//        return super.getValue(key)
//      }
//    }
  }
  private val mySecondaryOptionsComponent = JPanel(BorderLayout()).also {
    it.setBorder(BorderFactory.createEmptyBorder(10, 0, 0, 0))
  }
  protected var hasOverwriteOption: Boolean = true
  protected var updateChosenFile: (File) -> File = { it }
  protected var proposeChosenFile: () -> File = { File(defaultFileName) }

  protected val defaultFileName: String
    get() = if (myDocument == null) "document.gan" else myDocument.getFileName()


  override val title: String get() = pageTitle ?: ""

  fun tryChosenFile(file: File?) {
    validateFile(file).onSuccess {
      showError(null)
      preferences.put(PREF_SELECTED_FILE, fxFile.value?.absolutePath)
    }.onFailure {
      showError(it ?: "Something went wrong")
    }
  }

  protected open fun loadPreferences() {
    val oldFile = preferences.get(PREF_SELECTED_FILE, null)
    if (oldFile != null) {
      fxFile.value = updateChosenFile(File(oldFile))
    } else {
      fxFile.value = proposeChosenFile()
    }
  }

  override fun setActive(isActive: Boolean) {
    val optionGroups = this.optionGroups
    if (!isActive) {
      for (optionGroup in optionGroups) {
        optionGroup.commit()
      }
      if (allowMultipleChoice) {
        // TODO: restore fxFiles too?
      } else {
        fxFile.value?.let {
          preferences.put(PREF_SELECTED_FILE, it.absolutePath)
        }
      }
    } else {
      for (optionGroup in optionGroups) {
        optionGroup.lock()
      }

      fileFilter = createFileFilter()
      loadPreferences()
      resetCenterPane()
    }
  }

  protected open fun createSecondaryOptionsPanel(): Component? {
    return myOptionsBuilder.buildPlanePage(this.optionGroups.toTypedArray())
  }

  protected open fun createSecondaryOptionsPanelFx(): Node? = null

  protected abstract fun createFileFilter(): FileExtensionFilter?

  protected abstract val optionGroups: List<GPOptionGroup>

  protected open fun validateFile(file: File?): Result<File?, String?> {
    return basicValidateFile(file)
  }
}

class ChosenFile(val file: File, val fileOption: ObservableFiles, val chosenFiles: ObservableList<ChosenFile>, val validator: (File) -> Result<File?, String?>) {
  val isValid: Boolean get() = validator(file).isOk

  fun remove() {
    fileOption.set(fileOption.value.filter { it != file }, chosenFiles)
  }

}

class ListCellImpl: GPListCell<ChosenFile>() {
  override fun updateItem(item: ChosenFile?, empty: Boolean) {
    whenNotEmpty(item, empty) { chosenFile ->
      graphic = hbox {
        styleClasses.add("chosen-file-item")
        if (!chosenFile.isValid) {
          styleClasses.add("validation-error")
        }
        isSelected = this@ListCellImpl.isSelected
        label = chosenFile.file.name.asObservable()
        actions.add(GPAction.create("impex.fileChooserPage.action.remove") {
          FXThread.runLater(chosenFile::remove)
        }.also {
          it.putValue(GPAction.TEXT_DISPLAY, ContentDisplay.GRAPHIC_ONLY)
        })
      }
    }
  }
}

const val PREF_SELECTED_FILE: String = "selected_file"
